tests/rust/pipe-test/async_close/mod.rs

```diff
@@ -0,0 +1,199 @@
+fn async_close_pipe_multiple_times_2(libos: &mut LibOS, pipe_name: &str) -> Result<()> {
+    // Create a pipe.
+    let pipeqd: QDesc = match libos.create_pipe(pipe_name) {
+        Ok(pipeqd) => pipeqd,
+        Err(e) => anyhow::bail!("create_pipe() failed ({})", e),
+    };
+
+    // Succeed to issue asynchronous close.
+    let qt1: QToken = match libos.async_close(pipeqd) {
+        Ok(qt) => qt,
+        Err(e) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("async_close() failed ({})", e);
+        },
+    };
+
+    // Succeed to issue asynchronous close.
+    let qt2: QToken = match libos.async_close(pipeqd) {
+        Ok(qt) => qt,
+        Err(e) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("async_close() failed ({})", e);
+        },
+    };
+
+    // Poll once to ensure the async_close() coroutine runs and finishes the close.
+    match libos.wait(qt1, Some(Duration::from_micros(0))) {
+        Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_CLOSE && qr.qr_ret == 0 => {},
+        Ok(_) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("wait() should succeed with async_close()")
+        },
+        Err(_) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("wait() should succeed with async_close()")
+        },
+    }
+
+    // Poll once to ensure the async_close() coroutine runs and finishes the close.
+    match libos.wait(qt2, Some(Duration::from_micros(0))) {
+        Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_FAILED && qr.qr_ret == libc::EBADF => Ok(()),
+        Ok(_) => {
+            anyhow::bail!("wait() should succeed with async_close()")
+        },
+        Err(_) => {
+            anyhow::bail!("wait() should succeed with async_close()")
+        },
+    }
+}
+
+/// Attempts to asynchronously close the same pipe multiple times.
+fn async_close_pipe_multiple_times_3(libos: &mut LibOS, pipe_name: &str) -> Result<()> {
+    // Create a pipe.
+    let pipeqd: QDesc = match libos.create_pipe(pipe_name) {
+        Ok(pipeqd) => pipeqd,
+        Err(e) => anyhow::bail!("create_pipe() failed ({})", e),
+    };
+
+    // Succeed to issue asynchronous close.
+    let qt1: QToken = match libos.async_close(pipeqd) {
+        Ok(qt) => qt,
+        Err(e) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("async_close() failed ({})", e);
+        },
+    };
+
+    // Succeed to issue asynchronous close.
+    let qt2: QToken = match libos.async_close(pipeqd) {
+        Ok(qt) => qt,
+        Err(e) => {
+            println!("[ERROR] leaking pipeqd={:?}", pipeqd);
+            anyhow::bail!("async_close() failed ({})", e);
+        },
+    };
+
+    // Poll once to ensure the async_close() coroutine runs and finishes the close.
+    match libos.wait(qt2, Some(Duration::from_micros(0))) {
```

Comment:
Same here. We should not assume which close will succeed. In the future, the correct behavior should be that the second async_close fails and does not return a qt because we shouldn't schedule two async_close coroutines but we don't have that ability right now.

Reply:
Fixed.